vision/odlc_contour_filtering.py

```diff
@@ -0,0 +1,498 @@
+) -> bool:
+    """
+    Calculates the area of an upright bounding box around the given contour
+    and compares it to the rectangle (image) of the given dimensions
+
+    Parameters
+    ----------
+    contour : consts.Contour
+        The individual contour to be evaluated (as returned from cv2.findContours)
+    dims : tuple[int, int]
+        The dimensions of the image the contour is from
+        dim_height : int
+            The height of the image
+        dim_width : int
+            The width of the image
+    test_area_ratio : float = 10.0
+        The minimum acceptable ratio of image area to contour bounding box area
+
+    Returns
+    -------
+    acceptable_area_ratio : bool
+        Returns true if the bounding box area is less than the image area by a factor of
+        test_area_ratio or more
+    """
+    bounding_box_retval: tuple[int, int, int, int] = cv2.boundingRect(contour)
+    bounding_box: bbox.BoundingBox = bbox.BoundingBox(
+        bbox.tlwh_to_vertices(
+            bounding_box_retval[1],
+            bounding_box_retval[0],
+            bounding_box_retval[3] - bounding_box_retval[1],
+            bounding_box_retval[2] - bounding_box_retval[0],
+        ),
+        bbox.ObjectType.STD_OBJECT,
+    )
+
+    box_area: float = bounding_box.get_height() * bounding_box.get_width()
+    img_area: float = dims[0] * dims[1]
+
+    return box_area * test_area_ratio <= img_area
+
+
+def test_spikiness(contour: consts.Contour) -> bool:
+    """
+    Checks whether the contour has some points that are much farther away from the center of mass
+    than the rest of the points
+    basically real shapes will not have random points that are really far away from the rest
+    but a weird patch of grass or something without well defined edges will have crazy points
+    all over the place
+
+    Parameters
+    ----------
+    contour : consts.Contour
+        The individual contour to be evaluated (as returned from cv2.findContours)
+
+    Returns
+    -------
+    lacks_spikes : bool
+        True unless there is a statistical outlier point that is uniquely far away (or close ig)
+
+    References
+    -----
+    This function uses a concept called the "Image moment" to calculate the center of a given
+    contour. More information can be found here: https://en.wikipedia.org/wiki/Image_moment
+    """
+    moments: dict[str, float] = cv2.moments(contour)
+    # com is Center of Mass, com = (y_coord, x_coord)
+    # m01/m00 is the y coordinate of the center of the contour
+    # m10/m00 is the x coordinate of the center of the contour
+    com: NDArray[Shape["2"], Float32] = np.array(
+        ((moments["m01"] / moments["m00"]), (moments["m10"] / moments["m00"])), dtype=np.float64
+    )
+
+    # numpy array that will hold the distance of each point in the contour to the center of the
+    # contour
+    dists_com: NDArray[Shape["*"], Float32] = np.ndarray(contour.shape[0], Float32)
```

Comment:
This could be a 1 line comment if you made it more brief. Don't need to say its a numpy array since you have the type hint there.

Can start the comment as "Holds the distance..."